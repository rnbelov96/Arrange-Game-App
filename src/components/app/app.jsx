import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Greating from '../greating-screen/greating-screen';
import { AppActionCreators } from '@/reducer/app';
import {
  getCellsToClick, getDimention, getField, getFinishStatus, getStartStatus,
} from '@/reducer/app/selectors';
import GameScreen from '../game-screen/game-screen';

function App({
  isStarted,
  isFinished,
}) {
  return (
    <>
      {!isStarted ? <Greating /> : null}
      {!isFinished && isStarted ? (
        <GameScreen />
      ) : null}
    </>
  );
}

App.propTypes = {
  isStarted: PropTypes.bool.isRequired,
  isFinished: PropTypes.bool.isRequired,
};

function mapStateToProps(state) {
  return {
    field: getField(state),
    cellsToClick: getCellsToClick(state),
    dimention: getDimention(state),
    isStarted: getStartStatus(state),
    isFinished: getFinishStatus(state),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onClick: index => dispatch(AppActionCreators.moveCell(index)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
