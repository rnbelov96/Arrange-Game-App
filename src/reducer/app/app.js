import fieldCreator from '@/utils/field-creator';
import findCellsToClick from '@/utils/find-cells-to-click';

const initialState = {
  field: [],
  emptyCell: null,
  cellsToClick: [],
  inputValue: 'three',
  correctCellsArray: [],
  isStarted: false,
  isFinished: false,
};

const ActionTypes = {
  MOVE_CELL: 'MOVE_CELL',
  CHANGE_INPUT: 'CHANGE_INPUT',
  START_GAME: 'START_GAME',
};

const ActionCreators = {
  moveCell: data => ({
    type: ActionTypes.MOVE_CELL,
    payload: data,
  }),

  changeInput: data => ({ type: ActionTypes.CHANGE_INPUT, payload: data }),

  startGame: () => ({ type: ActionTypes.START_GAME }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.START_GAME:
      const { startField, correctCellsArray, emptyCell } = fieldCreator(
        state.inputValue,
      );
      return {
        ...state,
        isStarted: true,
        isFinished: false,
        field: startField,
        emptyCell,
        cellsToClick: findCellsToClick(emptyCell, state.inputValue),
        correctCellsArray,
      };

    case ActionTypes.MOVE_CELL:
      const modifiedField = [...state.field];
      if (state.cellsToClick.includes(action.payload)) {
        modifiedField[state.emptyCell] = modifiedField[action.payload];
        modifiedField[action.payload] = null;
      }
      if (
        !modifiedField[modifiedField.length - 1]
        && JSON.stringify(modifiedField)
          === JSON.stringify(state.correctCellsArray)
      ) {
        return {
          ...state,
          isFinished: true,
          isStarted: false,
          field: modifiedField,
          emptyCell: modifiedField.findIndex(el => !el),
          cellsToClick: [],
        };
      }
      return {
        ...state,
        field: modifiedField,
        emptyCell: modifiedField.findIndex(el => !el),
        cellsToClick: findCellsToClick(
          modifiedField.findIndex(el => !el),
          state.inputValue,
        ),
      };

    case ActionTypes.CHANGE_INPUT:
      return {
        ...state,
        inputValue: action.payload,
      };
    default:
      return state;
  }
};

export {
  reducer as appReducer,
  ActionCreators as AppActionCreators,
  ActionTypes as DataActionTypes,
};
