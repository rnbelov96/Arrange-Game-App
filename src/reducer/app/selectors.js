export const getField = state => state.app.field;

export const getCellsToClick = state => state.app.cellsToClick;

export const getDimention = state => state.app.inputValue;

export const getStartStatus = state => state.app.isStarted;

export const getFinishStatus = state => state.app.isFinished;
