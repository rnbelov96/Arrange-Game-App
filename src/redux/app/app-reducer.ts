import { correctFieldArrays } from '@/constants';
import { DimentionType } from '@/types/general-types';
import {
  AppActionConstType, AppActionType, AppInitialStateType, ChangeInputActionType, MoveCellActionType, StartGameActionType,
} from '@/types/redux/app-reducer';
import fieldCreator from '@/utils/field-creator';
import findCellsToClick from '@/utils/find-cells-to-click';

const initialState: AppInitialStateType = {
  field: [],
  emptyCell: null,
  cellsToClick: [],
  inputValue: 'three',
  isStarted: false,
  isFinished: false,
};

const ActionTypes: AppActionConstType = {
  MOVE_CELL: 'MOVE_CELL',
  CHANGE_INPUT: 'CHANGE_INPUT',
  START_GAME: 'START_GAME',
};

const ActionCreators = {
  moveCell: (data: number): MoveCellActionType => ({
    type: ActionTypes.MOVE_CELL,
    payload: data,
  }),

  changeInput: (data: DimentionType): ChangeInputActionType => ({
    type: ActionTypes.CHANGE_INPUT, payload: data,
  }),

  startGame: (): StartGameActionType => ({ type: ActionTypes.START_GAME }),
};

const reducer = (state: AppInitialStateType = initialState, action: AppActionType) => {
  switch (action.type) {
    case ActionTypes.START_GAME:
      const { startField, emptyCell } = fieldCreator(
        state.inputValue,
      );
      return {
        ...state,
        isStarted: true,
        isFinished: false,
        field: startField,
        emptyCell,
        cellsToClick: findCellsToClick(emptyCell, state.inputValue),
      };

    case ActionTypes.MOVE_CELL:
      const modifiedField = [...state.field];
      if (state.cellsToClick.includes(action.payload)) {
        if (state.emptyCell !== null) {
          modifiedField[state.emptyCell] = modifiedField[action.payload];
        }
        modifiedField[action.payload] = null;
      }
      if (
        !modifiedField[modifiedField.length - 1]
        && JSON.stringify(modifiedField)
          === JSON.stringify(correctFieldArrays[state.inputValue])
      ) {
        return {
          ...state,
          isFinished: true,
          isStarted: false,
          field: modifiedField,
          emptyCell: modifiedField.findIndex(el => !el),
          cellsToClick: [],
        };
      }
      return {
        ...state,
        field: modifiedField,
        emptyCell: modifiedField.findIndex(el => !el),
        cellsToClick: findCellsToClick(
          modifiedField.findIndex(el => !el),
          state.inputValue,
        ),
      };

    case ActionTypes.CHANGE_INPUT:
      return {
        ...state,
        inputValue: action.payload,
      };
    default:
      return state;
  }
};

export {
  reducer as appReducer,
  ActionCreators as AppActionCreators,
  ActionTypes as DataActionTypes,
};
