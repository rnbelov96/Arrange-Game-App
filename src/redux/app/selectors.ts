import { FullStateType } from '@/types/general-types';

export const getField = (state: FullStateType) => state.app.field;

export const getCellsToClick = (state: FullStateType) => state.app.cellsToClick;

export const getDimention = (state: FullStateType) => state.app.inputValue;

export const getStartStatus = (state: FullStateType) => state.app.isStarted;

export const getFinishStatus = (state: FullStateType) => state.app.isFinished;
