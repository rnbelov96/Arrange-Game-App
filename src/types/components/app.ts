export type AppPropsType = {
  isFinished: boolean,
  isStarted: boolean,
};
