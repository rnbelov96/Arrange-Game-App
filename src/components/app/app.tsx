import React from 'react';
import { GameScreen } from 'components/game-screen';
import { GreatingScreen } from 'components/greating-screen';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { AppActionCreators } from '@/redux/app';
import {
  getCellsToClick, getDimention, getField, getFinishStatus, getStartStatus,
} from '@/redux/app/selectors';
import { AppPropsType } from '@/types/components/app';
import { AppActionType } from '@/types/redux/app-reducer';
import { FullStateType } from '@/types/general-types';

export const PureApp: React.FunctionComponent<AppPropsType> = ({
  isStarted,
  isFinished,
}: AppPropsType) => (
  <>
    {!isStarted ? <GreatingScreen /> : null}
    {!isFinished && isStarted ? (
      <GameScreen />
    ) : null}
  </>
);

const mapStateToProps = (state: FullStateType) => ({
  field: getField(state),
  cellsToClick: getCellsToClick(state),
  dimention: getDimention(state),
  isStarted: getStartStatus(state),
  isFinished: getFinishStatus(state),
});

const mapDispatchToProps = (dispatch: Dispatch<AppActionType>) => ({
  onClick: (index:number) => dispatch(AppActionCreators.moveCell(index)),
});

export const App = connect(mapStateToProps, mapDispatchToProps)(PureApp);
