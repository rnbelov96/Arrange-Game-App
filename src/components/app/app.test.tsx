import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { App } from './app';

const mockStore = configureStore([]);

describe('App', () => {
  test('should render greating screen with greating message', () => {
    const store = mockStore({
      app: {
        field: [null, 7, 6, 8, 4, 1, 3, 5, 2],
        emptyCell: 0,
        cellsToClick: [3, 1],
        inputValue: 'three',
        isStarted: false,
        isFinished: false,
      },
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <App />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render greating screen with winning message', () => {
    const store = mockStore({
      app: {
        field: [null, 7, 6, 8, 4, 1, 3, 5, 2],
        emptyCell: 0,
        cellsToClick: [3, 1],
        inputValue: 'three',
        isStarted: false,
        isFinished: true,
      },
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <App />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render game screen', () => {
    const store = mockStore({
      app: {
        field: [null, 7, 6, 8, 4, 1, 3, 5, 2],
        emptyCell: 0,
        cellsToClick: [3, 1],
        inputValue: 'three',
        isStarted: true,
        isFinished: false,
      },
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <App />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
