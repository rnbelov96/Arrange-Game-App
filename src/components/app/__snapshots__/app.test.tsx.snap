// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render game screen 1`] = `
<div
  className="sc-AykKC gEhdZJ"
>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  />
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    7
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    6
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    8
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    4
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    1
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    3
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    5
  </div>
  <div
    className="sc-AykKD fKtPur"
    onClick={[Function]}
    style={
      Object {
        "width": "33%",
      }
    }
  >
    2
  </div>
</div>
`;

exports[`App should render greating screen with greating message 1`] = `
<div
  className="sc-AykKE btosjx"
>
  <h1
    className="sc-AykKF ikUiPZ"
  >
    Arrange Game
  </h1>
  <h2
    className="sc-AykKG hXQWYg"
  >
    Chose dimention:
  </h2>
  <select
    className="sc-AykKH jcCLzn"
    onChange={[Function]}
    value="three"
  >
    <option
      value="three"
    >
      three
    </option>
    <option
      value="four"
    >
      four
    </option>
    <option
      value="five"
    >
      five
    </option>
  </select>
  <button
    className="sc-AykKI klDJOA"
    onClick={[Function]}
    type="button"
  >
    Start
  </button>
</div>
`;

exports[`App should render greating screen with winning message 1`] = `
<div
  className="sc-AykKE btosjx"
>
  <h1
    className="sc-AykKF ikUiPZ"
  >
    WINNER!!!
  </h1>
  <h2
    className="sc-AykKG hXQWYg"
  >
    Chose dimention:
  </h2>
  <select
    className="sc-AykKH jcCLzn"
    onChange={[Function]}
    value="three"
  >
    <option
      value="three"
    >
      three
    </option>
    <option
      value="four"
    >
      four
    </option>
    <option
      value="five"
    >
      five
    </option>
  </select>
  <button
    className="sc-AykKI klDJOA"
    onClick={[Function]}
    type="button"
  >
    Start
  </button>
</div>
`;
